### DISCUSSION TITLE
Buffered Ni-Daq acquisition with multiple in- and outputs

### DISCUSSION THREAD

**Discussion by mgraupe** [bonsai-rx/bonsai#1326] [Q&A]
Hello, 

I am implementing our experiment in Bonsai and I am loving it. The experiment comprises the simultaneous acquisition and control of two cameras and several digital and analog in and output lines recorded through a Ni-Daq board. My current state of implementation is that the cameras (and frame timing information) are recorded for a defined period  after a key-press (see attached screenshot of the bonsai sketch). I have two unrelated questions : 

1. I would like to display whether the cameras are in preview or in an active acquisition mode. In other words, depending on whether the timer is active or not, a display should toggle between "preview" or "recording". How can this be implemented? 

2. For a predefined period of time and sampling rate (e.g. 60 sec, 20 kHz), I want to record and send out multiple signals from the Ni-Daq board. Importantly, the timing relationship should be fixed between them. In my current control software ACQ4, this is implemented as a buffered operation running on the ni-daq board where the participating channels are simultaneously sampled. I have not found an example for a related Bonsai sketch and I would appreciate any hints or examples to start from. 

Thank you very much in advance. 

Best regards,
Michael 

![Screenshot 2023-04-19 103527](https://user-images.githubusercontent.com/8803795/233064707-b6dc3c96-3eac-4d2b-9423-a15e570771f7.png)

**Comment by bruno-f-cruz** [comment]
You can do the same thing in bonsai by defining multiple channels when you define your task (`AnalogOutput`) and sending a matrix with the number of rows equal to the number of channels (eg. by merging two `FunctionGenerator` nodes). This will lead to all the analog outputs playing back the matrix synchronously across all the channels in that task.

**Reply by mgraupe** [reply]
Dear Bruno, 

Thank you for your prompt reply. 
I suppose what you describe also works for digital outputs. ?
And how can that be combined with data acquisition - e.g. digital input - recorded synchronously with the outputs? 
Do you know of examples using the Ni-Daq package where I can find inspiration? 

Do you have an idea of how to implement my question 1) above? 

Thanks,
Michael

**Reply by bruno-f-cruz** [reply]
"I suppose what you describe also works for digital outputs. ?"
As long as your hardware is capable of buffering the channels involved in the task it will.

"And how can that be combined with data acquisition - e.g. digital input - recorded synchronously with the outputs?"
You probably want to use the `SignalSource` property and use an event from one task to trigger the acquisition of the second one.  Does your hardware provide a signal source associated with the sampling of the digital input or output? If so, feed this event in a form of a string (like you would in any other software using the daqmx API) to this property. 

"Do you know of examples using the Ni-Daq package where I can find inspiration?"
The documentation of the package is likely the best place to start. Right-click on a node of the package, followed by"view help". This will get you right onto the website.

"Do you have an idea of how to implement my question 1) above?"
Can you open a new discussion with this one?

Best,
B
